libs/model/src/contest/Contests.projection.ts

```diff
@@ -0,0 +1,88 @@
+import { Projection, schemas } from '@hicommonwealth/core';
+import { models } from '../database';
+import { mustExist } from '../middleware/guards';
+
+const inputs = {
+  RecurringContestManagerDeployed:
+    schemas.events.RecurringContestManagerDeployed,
+  OneOffContestManagerDeployed: schemas.events.OneOffContestManagerDeployed,
+  ContestStarted: schemas.events.ContestStarted,
+  ContestContentAdded: schemas.events.ContestContentAdded,
+  ContestContentUpvoted: schemas.events.ContestContentUpvoted,
+  ContestWinnersRecorded: schemas.events.ContestWinnersRecorded,
+};
+
+export const Contests: Projection<typeof inputs> = () => ({
+  inputs,
+  body: {
+    RecurringContestManagerDeployed: async ({ payload }) => {
+      const community = await models.Community.findOne({
+        where: { namespace: payload.namespace },
+        raw: true,
+      });
+      if (
+        mustExist(`Community with namespace: ${payload.namespace}`, community)
+      )
+        await models.ContestManager.upsert({
```

Comment:
Why are we using `upsert` in all of these projections? Isn't the communityId, interval, etc of a specific contest manager immutable? Feels like this could lead to accidentally updating contest manager records due to bugs.

Reply:
ha, I responded to this comment two days ago and GH lost it 🤷 

This is a simple way to achieve idempotent projections... but just in case, if we receive duplicate events or out-of-order events, this specific projection will reflect those facts. 

We can find other patterns to achieve the same results, noting that every projection is different and might require other measures.

Reply:
This only makes the projection idempotent with respect to the primary key (or unique constraint) of the table. This means if we emit 2 events with the same primary key but different metadata, we update the record thereby overwriting the original values. So rather than receiving an error for a duplicate event (which we can safely ignore silently) we introduce an avenue for a bug that causes data corruption and leaves us unaware of its presence entirely until some downstream effect manifests itself.

Reply:
we have to look at these issues case by case... can you think of an example where two contest managers with the same key and different metadata are actually allowed in the system?

Reply:
I think you are trying to conceptualize projections are another layer of system testing, but in order to keep it simple, projections are better conceptualized as "dumb" caches of facts that the system allowed to happen... we can test corruption at a different level (monitors, analytics)

Reply:
No but that doesn't negate the point that we could technically introduce a bug where data is mismatched. If we don't do upsert then when this happens we know about it because we get a duplicate key constraint error. If we do upsert we overwrite the data and the bug remains.

Of course this is an unlikely bug but I just don't see why we would open ourselves up to it.

Reply:
Right so if we test for corruption at different levels shouldn't we assume that there is no corruption at the projection level...

Reply:
I see your point, and in this case (a genesis event) insert makes more sense... for the rest of events, we can debate upserts or updates ... the point is that if we get one out-of-order event, it will be upserted and the projection will fail the next insert... some projections will even allow concurrent updates (just patches)

the only way to make this strict based on inserts/updates is by ensuring events will be 100% ordered, and that can be achieved too

Reply:
last commit using "create"... we'll see how it goes with the ordered outbox 🙏

Reply:
I don't think `insert` should be limited to just the genesis event but all events as long as `update` is not a strictly required operation. We should discuss this in huddle.